sdk/core/Azure.Core/src/Response.cs

```diff
@@ -154,5 +138,150 @@ internal static void DisposeStreamIfNotBuffered(ref Stream? stream)
                 stream = null;
             }
         }
+
+        /// <summary>
+        /// TBD.
+        /// </summary>
+        /// <param name="cancellationToken"></param>
+        /// <returns></returns>
+        /// <exception cref="NotImplementedException"></exception>
+        public override BinaryData ReadContent(CancellationToken cancellationToken = default)
+        {
+            // Derived types should provide an implementation that allows caching
+            // to improve performance.
+            if (ContentStream is null)
+            {
+                return s_EmptyBinaryData;
+            }
+
+            if (ContentStream is MemoryStream memoryStream)
+            {
+                return memoryStream.TryGetBuffer(out ArraySegment<byte> segment) ?
+                    new BinaryData(segment.AsMemory()) :
+                    new BinaryData(memoryStream.ToArray());
+            }
+
+            BufferedContentStream bufferStream = new();
+
+            Stream? contentStream = ContentStream;
+            contentStream.CopyTo(bufferStream, cancellationToken);
+            contentStream.Dispose();
+
+            bufferStream.Position = 0;
+            ContentStream = bufferStream;
+
+            BinaryData content = BinaryData.FromStream(bufferStream);
+            bufferStream.Position = 0;
+
+            return content;
+        }
+
+        /// <summary>
+        /// TBD.
+        /// </summary>
+        /// <param name="cancellationToken"></param>
+        /// <returns></returns>
+        /// <exception cref="NotImplementedException"></exception>
+        public override async ValueTask<BinaryData> ReadContentAsync(CancellationToken cancellationToken = default)
+        {
+            // Derived types should provide an implementation that allows caching
+            // to improve performance.
+            if (ContentStream is null)
+            {
+                return s_EmptyBinaryData;
+            }
+
+            if (ContentStream is MemoryStream memoryStream)
+            {
+                return memoryStream.TryGetBuffer(out ArraySegment<byte> segment) ?
+                    new BinaryData(segment.AsMemory()) :
+                    new BinaryData(memoryStream.ToArray());
+            }
+
+            BufferedContentStream bufferStream = new();
+
+            Stream? contentStream = ContentStream;
+            await contentStream.CopyToAsync(bufferStream, cancellationToken).ConfigureAwait(false);
```

Comment:
This appears to be the only line difference from the sync version above. Could we de-dupe into a syncAsync version?

Reply:
Yeah, good call.  I was planning to refactor that when I came back to it.  This comment can serve as my reminder.

Reply:
Addressed in https://github.com/Azure/azure-sdk-for-net/pull/42234